vernac/vernacentries.mli

```diff
@@ -14,11 +14,21 @@ val dump_global : Libnames.reference or_by_notation -> unit
 val vernac_require :
   Libnames.reference option -> bool option -> Libnames.reference list -> unit
 
+type interp_state = { (* TODO: inline records in OCaml 4.03 *)
+  system  : States.state;        (* summary + libstack *)
+  proof   : Proof_global.state;  (* proof state *)
+  shallow : bool                 (* is the state trimmed down (libstack) *)
+}
+
+val freeze_interp_state : Summary.marshallable -> interp_state
+val unfreeze_interp_state : interp_state -> unit
```

Comment:
Something is wrong here.  The `invalidate_cache` is not exported, hence nobody can invalidate it unless it uses `with_fail`. (and, more importantly, interp does not invalidate if an exception is raised).

As a consequence interpreting is unsafe when the interpreted command raises an exception after having tampered with the imperative state of coq. I mean, a call to unfreeze may result in a no-op.

IMO the right fix is to have interp call invalidate on exceptions

Reply:
The cache is private, thus users of the API have no business observing it. For them, interpretation is always functional so they don't care. You are right about exceptions, even if in most cases the cache is not updated so this is not a problem in practice, the cache will be invalid.